doc/spec_TransferElement.dox

```diff
@@ -213,9 +232,50 @@ This documnent is currently still **INCOMPLETE**!
+    - 12.1.3 After a TransferElement is added to a TransferGroup, TransferElement::replaceTransferElement() is called for each element in the group, passing each internal element in the group as a replacement.
+    - 12.1.4 TransferElement::replaceTransferElement() uses the given replacement TransferElement for an optimisation, if possible. Depending on the protocol, this can mean:
+      - \anchor transferElement_B_12_1_4_1 12.1.4.1 An internal element is replaced 1:1 by the replacement, because it transfers the same data (cf. \ref transferElement_B_12_1_5 "12.1.5"). This is a common case for decorator-like implementations.
+      - 12.1.4.2 The given replacement can be extended such that the data of an internal element is included in the transfer; then the internal element can be replaced by the now-extended replacement.
+      - 12.1.4.3 If such replacement takes place, the behaviour visible to the application of the TransferElement does not change (exception: VersionNumbers in read opertions can reveal now-consistent data, see 12.1.1).
+      - 12.1.4.4 If no such optimisation is possible, the function simply does nothing.
+    - \anchor transferElement_B_12_1_5 12.1.5 TransferElement::mayReplaceOther() returns true, if the given TransferElement behaves identical to the one it is called on in every aspect, i.e. it accesses the same data on the same device, has the same UserType and the same AccessModeFlags, and (if applicable) uses the same data transformations (like fixed point conversion). It returns false otherwise. This is used by implementations of \ref transferElement_B_12_1_4_1 "12.1.4.1".
+  - 12.2 Adding TransferElements which cannot merge their transfers to the same TransferGroup is always allowed.
+  - 12.3 If TransferElements have been added to a TransferGroup, individual read/write operations of the TransferElements are no longer allowed and will throw a ChimeraTK::logic_error. Read/write operations are then only possible through the TransferGroup.
+  - 12.4 A TransferGroup is becomes read-only (cf. TransferGroup::isReadOnly()), if at least one of the member TransferElements is read-only.
+  - 12.5 Adding the same TransferElement twice is not allowed (logic_error).
+  - 12.6 Adding one TransferElement to two different TransferGroups is not allowed (logic_error).
+  - 12.7 Adding TransferElements with AccessMode::wait_for_new_data is not allowed (logic_error).
+  - 12.8 Adding a copy of the a TransferElement which is already in the group (returned by two different calls to Device::getZzzRegisterAccessor() resp. DeviceBackend::getRegisterAccessor() with the same arguments) will replace one of the identical TransferElement with a decorator which copies the data from the other.
+    - 12.8.1 This avoids an unnecessary transfer of identical data.
+    - 12.8.2 This is only possible, if the TransferElement is added by passing a TransferElementAbstractor to TransferGroup::addAccessor().
```

Comment:
You mean why it is only possible when an abstractor is used? Because you need some kind of "holder" so you can replace the highest level. This is a logical consequence of the interface. The behaviour is influenced by it so it needs to be mentioned.
"Implementation detail" means it can be done differently without anyone noticing. This is not the case here.

Reply:
Ah, the highest element is exchanged. Then it is clear why this can only be done when you use the abstractor.

But why is this treated differently to another TE without abstractor with the same target being added, where they can share the same low level TE?

Ah, I see, this also works if they cannot share the transfer because they don't have low level TEs and do not support replacing each other in the low level. This should go to a comment.

Still I don't see why you would want to add different TEs to the same data multiple times.

Reply:
> Still I don't see why you would want to add different TEs to the same data multiple times.

It can happen even somewhat indirectly if two logical name mapping registers redirect to the same register. It's not a super important feature and I probably would not implement it any more today, but it has some usefulness and does not overcomplicate things too much, so I would keep it in.

It is actually not treated differently, there is something missing in the spec: A copy decorator is also used when replacing a transfer element on the inner levels 1:1, unless the implementation can guarantee that all peers accessing the same low-level element are keeping its application data buffer intact (i.e. they do not swap out the buffer). For that reason, NDRegisterAccessorDecorator::replaceTransferElement() creates a copy decorator, because it swaps the data even itself. I just found on the other hand that the LMNBackendBitAccessor::replaceTransferElement() does not create a copy decorator, which is a bug. So far it goes unnoticed, because usually the peers are only other LMNBackendBitAccessors, which do not swap out the data. It would go wrong, if the target accessor of a bit accessor would be used as a target of a decorator which swaps out the data, and by chance the decorators postRead is called first. Then the bit accessors would see wrong data. (Might still go unnoticed btw., because data would simply be outdated... Hard to catch!)

I will add a comment.